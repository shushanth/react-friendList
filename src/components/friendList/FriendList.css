:local(.friendList) {
  padding: 0;
  margin: 0;
  max-height: 100%;
  overflow: hidden;
  display: flex;
  flex-direction: column;
}

:local(.friendListItems) {
  display: inherit;
}

:local(.friendListItem) {
  background-color: white;
  border-bottom: 1px solid #e3e3e3;
  display: inherit;
  list-style: none;
  padding: 20px;
  width: 100%;
}

:local(.friendInfos) {
  flex: 1 0 auto;
}

:local(.friendInfos span) {
  font-weight: bold;
}

:local(.btnAction),
:local(.btnAction):active,
:local(.btnAction):focus,
:local(.btnAction):hover {
  margin-right: 5px;
  color: #5c75b0;
}

button:focus {
  outline: 0 !important;
}
