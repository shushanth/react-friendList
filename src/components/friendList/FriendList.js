import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import { GridRow, GridColumn } from "../../shared/components";
import { uuid } from "../../utils";

import styles from "./FriendList.css";

const FriendList = ({ friends, selectedPage, pageSize, actions }) => {
  let friendsList = [...friends];
  let filteredFriends = [];
  for (let i = 0; i <= selectedPage - 1; i++) {
    filteredFriends = friendsList.splice(0, pageSize);
  }
  const { starFriend, deleteFriend, pageMove } = actions;
  const onDeletFriend = id => {
    deleteFriend(id);
    setTimeout(() => pageMove());
  };
  return (
    <div className={classnames(styles.friendList)}>
      {filteredFriends.map(list => {
        return (
          <div className={classnames(styles.friendListItems)} key={uuid()}>
            <GridRow rowData={list}>
              {rowData => {
                return (
                  <div className={classnames(styles.friendListItem)}>
                    <GridColumn cell={rowData.name}>
                      {cellData => {
                        return (
                          <div className={classnames(styles.friendInfos)}>
                            <div>
                              <span>{rowData.name}</span>
                            </div>
                            <div>
                              <small>{rowData.gender}</small>
                            </div>
                            <div>
                              <small>xx friends in common</small>
                            </div>
                          </div>
                        );
                      }}
                    </GridColumn>
                    <GridColumn>
                      {cellData => {
                        return (
                          <div className={classnames(styles.friendActions)}>
                            <button
                              className={`btn btn-default ${styles.btnAction}`}
                              onClick={() => starFriend(list.id)}
                            >
                              <i
                                className={classnames("fa", {
                                  "fa-star": rowData.starred,
                                  "fa-star-o": !rowData.starred
                                })}
                              />
                            </button>
                            <button
                              className={`btn btn-default ${styles.btnAction}`}
                              onClick={() => onDeletFriend(list.id)}
                            >
                              <i className="fa fa-trash" />
                            </button>
                          </div>
                        );
                      }}
                    </GridColumn>
                  </div>
                );
              }}
            </GridRow>
          </div>
        );
      })}
    </div>
  );
};

FriendList.PropTypes = {
  lists: PropTypes.arrayOf(PropTypes.object),
  starFriend: PropTypes.func,
  deleteFriend: PropTypes.func
};

export default FriendList;
