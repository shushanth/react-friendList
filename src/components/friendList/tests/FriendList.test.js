import React from "react";
import { mount, shallow } from "enzyme";

import FriendList from "../FriendList";

describe("Component: <FriendList/>", () => {
  let FriendListWapper;
  let FriendListMountWapper;
  const friends = [
    {
      id: 1,
      name: "Theodore Roosevelt",
      gender: "Male",
      starred: true
    },
    {
      id: 2,
      name: "Abraham Lincoln",
      gender: "Female",
      starred: false
    },
    {
      id: 3,
      name: "George Washington",
      gender: "Male",
      starred: false
    }
  ];
  const actions = {
    starFriend: jest.fn(),
    deleteFriend: jest.fn(),
    pageMove: jest.fn()
  };
  beforeAll(() => {
    FriendListWapper = shallow(
      <FriendList friends={friends} actions={actions} />
    );
    FriendListMountWapper = mount(
      <FriendList
        selectedPage={1}
        pageSize={2}
        friends={friends}
        actions={actions}
      />
    );
  });

  it("should render without crashing", () => {
    expect(FriendListWapper.exists()).toBeTruthy();
    expect(FriendListWapper).toMatchSnapshot();
  });

  it("should have list count same as page size", () => {
    expect(FriendListMountWapper.find("GridRow").length).toEqual(2);
  });

  it("should have list columns same as per the friends", () => {
    let firstFriendList = FriendListMountWapper.find("GridRow").at(0);
    let secondFriendList = FriendListMountWapper.find("GridRow").at(1);
    expect(firstFriendList.find("GridColumn").length).toEqual(2);
    expect(secondFriendList.find("GridColumn").length).toEqual(2);
  });

  it("should match columns values as per the friends list", () => {
    let firstFriend = friends[0];
    let secondFriend = friends[1];
    let firstFriendList = FriendListMountWapper.find("GridRow").at(0);
    let secondFriendList = FriendListMountWapper.find("GridRow").at(1);
    let firstFriendListColumn = firstFriendList.find("GridColumn").at(0);
    let secondFriendListColumn = secondFriendList.find("GridColumn").at(0);
    expect(firstFriendListColumn.find("span").text()).toEqual(firstFriend.name);
    expect(
      firstFriendListColumn
        .find("small")
        .at(0)
        .text()
    ).toEqual(firstFriend.gender);
    expect(secondFriendListColumn.find("span").text()).toEqual(
      secondFriend.name
    );
    expect(
      secondFriendListColumn
        .find("small")
        .at(0)
        .text()
    ).toEqual(secondFriend.gender);
  });

  describe("method: deleteFriend", () => {
    it("should invoke the actions delete method on first list item", () => {
      let deleteFriendListFirstBtn = FriendListMountWapper.find("GridRow")
        .at(0)
        .find("GridColumn")
        .at(1)
        .find("button")
        .at(1);
      deleteFriendListFirstBtn.simulate("click");
      expect(actions.deleteFriend).toHaveBeenCalled();
    });

    it("should invoke the actions delete method on second list item", () => {
      let deleteFriendListSecondBtn = FriendListMountWapper.find("GridRow")
        .at(1)
        .find("GridColumn")
        .at(1)
        .find("button")
        .at(1);
      deleteFriendListSecondBtn.simulate("click");
      expect(actions.deleteFriend).toHaveBeenCalled();
    });
  });

  describe("starredFriend", () => {
    it("should invoke actions starFriend method, click of star friend button in first list", () => {
      let starFriendListFirstBtn = FriendListMountWapper.find("GridRow")
        .at(0)
        .find("GridColumn")
        .at(1)
        .find("button")
        .at(0);
      starFriendListFirstBtn.simulate("click");
      expect(actions.starFriend).toHaveBeenCalled();
    });
    it("should invoke actions starFriend method, click of star friend button in second list", () => {
      let starFriendListSecondBtn = FriendListMountWapper.find("GridRow")
        .at(1)
        .find("GridColumn")
        .at(1)
        .find("button")
        .at(0);
      starFriendListSecondBtn.simulate("click");
      expect(actions.starFriend).toHaveBeenCalled();
    });
  });
});
