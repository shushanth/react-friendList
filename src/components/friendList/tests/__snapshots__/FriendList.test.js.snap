exports[`Component: <FriendList/> should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FriendList
    actions={
        Object {
            "deleteFriend": [Function],
            "pageMove": [Function],
            "starFriend": [Function],
          }
    }
    friends={
        Array [
            Object {
              "gender": "Male",
              "id": 1,
              "name": "Theodore Roosevelt",
              "starred": true,
            },
            Object {
              "gender": "Female",
              "id": 2,
              "name": "Abraham Lincoln",
              "starred": false,
            },
            Object {
              "gender": "Male",
              "id": 3,
              "name": "George Washington",
              "starred": false,
            },
          ]
    } />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
